File: 954-maximum-sum-circular-subarray/maximum-sum-circular-subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        total_sum = 0

        c_max = 0
        g_max = nums[0]

        current_min = 0
        global_min = nums[0]

        for i in nums:
            total_sum += i

            c_max = max(i,c_max+i)
            g_max = max(g_max,c_max)

            current_min = min(i, current_min + i)
            global_min = min(global_min, current_min)
        
        if g_max < 0:
            return g_max
    
        return max(g_max,total_sum-global_min)

File: 954-maximum-sum-circular-subarray/maximum-sum-circular-subarray.java
class Solution {
    public int maxSubarraySumCircular(int[] nums) {
        int cursum1 = 0;
        int cursum2 = 0;
        int totalSum = 0;
        int res1 = nums[0];
        int res2 = nums[0];
        for(int i:nums){
            totalSum += i;
            if(cursum1<0) cursum1=0;
            cursum1+=i;
            res1 = Math.max(res1,cursum1);
        }
        for(int i:nums){
            if(cursum2>0) cursum2=0;
            cursum2+=i;
            res2 = Math.min(res2,cursum2);
        }
        if(totalSum==res2) return res1;
        return Math.max(res1,totalSum-res2);
    }
}